clippy_lints/src/rebind_fn_arg_as_mut.rs

```diff
@@ -0,0 +1,117 @@
+use crate::utils::{snippet, span_lint_and_then};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::intravisit::{FnKind, NestedVisitorMap, Visitor};
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::hir::map::Map;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{symbol::Ident, Span};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for function arguments declared as not mutable and
+    /// later rebound as mutable.
+    ///
+    /// **Why is this bad?** It can be easily improved by just declaring the function
+    /// argument as mutable and removing the unnecessary assignment.
+    ///
+    /// **Known problems:** The function argument might have been shadowed by another
+    /// value before the assignment.
```

Comment:
I'm really not sure about this. Is there a simple way to prevent this? If not, we should probably mark this lint as pedantic, right?